### ISSUE TITLE
Raspberry Pi 4 - kernel panics

### ISSUE THREAD

**Issue by hikikones** [rapier1/web10g#11]
Hello. I am trying to make web10g work on a RPI4, but am experiencing kernel panics whenever I load the web10g kernel module and start logging. I will explain how I have installed web10g first. I am a bit new to Linux in general and C, so bear with me.

I am using [Raspbian Buster Lite](https://www.raspberrypi.org/downloads/raspbian/) for the OS for RPI4. I have then cloned this repo, and checked out the `kis-0.13-5.5` tag since linux mainline kernel 5.5.0 now supports the latest Raspberry Pi 4. Then I did a [local build](https://www.raspberrypi.org/documentation/linux/kernel/building.md) as specified by the raspberry pi guide. But in addition, I downloaded the [bcm2711_defconfig](https://github.com/raspberrypi/linux/tree/rpi-5.5.y/arch/arm/configs) needed for RPI4, and added the line `CONFIG_TCP_ESTATS=y` to it for web10g feature. During the initial build, the following errors appeared:

```
ERROR: "tcp_estats_update_finish_segrecv" [net/ipv6/ipv6.ko] unde
fined!
ERROR: "tcp_estats_create" [net/ipv6/ipv6.ko] undefined!
ERROR: "tcp_estats_update_segrecv" [net/ipv6/ipv6.ko] undefined! 
scripts/Makefile.modpost:103: recipe for target 'modules-modpost'
failed
make[1]: *** [modules-modpost] Error 1
Makefile:1301: recipe for target 'modules' failed
make: *** [modules] Error 2
```

I did not manage to fix these errors, but since I am not going to need ipv6, I simply commented out the function calls and then the build succeeded. Then I built the kernel module (dlkm), which resulted in `WARNING: "__aeabi_ldivmod" undefined`. To fix this, I removed the `u64` division cast as follows:

```
--- web10g-dlkm/tcp_estats_mib_var.c
+++ web10g-dlkm-arm/tcp_estats_mib_var.c
@@ -99,8 +99,8 @@
 static void read_StartTimeStamp(void *buf, struct tcp_estats *stats,
 				struct tcp_estats_var *vp)
 {
-	u64 msecs = (u64)stats->start_tv.tv_sec * 1000000; /* convert to msecs */
-	msecs = msecs + (u64)stats->start_tv.tv_nsec/1000; /* convert nsec to msec */
+	u64 msecs = stats->start_tv.tv_sec * 1000000; /* convert to msecs */
+	msecs = msecs + stats->start_tv.tv_nsec/1000; /* convert nsec to msec */
 	memcpy(buf, &msecs, 8);
 }
```

After that, the kernel module (dlkm) was built successfully. The web10g userland tools was built successfully with no modifications.

With everything installed, I enabled web10g with `sudo modprobe tcp_estats_nl` followed by `sudo sysctl -w net.ipv4.tcp_estats=127`. No kernel panics yet. However, if I now ssh into (or out of) the web10g computer from another computer, a kernel panic occurs. I can't find the log for the panic in either `/var/log` og `/var/crash`, but have an [image](https://imgur.com/a/DIRX70T) if that is of any help.

I would appreciate some help for this problem. Is there something I am doing very wrong? Or is there something more specific I need ebto do for w10g to work on ARM architecture?

**Comment by rapier1** [comment]
I have to admit that I haven't built Web10g for the raspberry pi. Part of the problem being that I don't have access to any of the raspberry pi hardware through work and I only have a CuBox at home. Let me see if anyone else here at work has a pi I can borrow for a few days to try this out. 

Did you build the kernel directly on the Pi or did you build it in a cross compiling environment?

**Comment by hikikones** [comment]
That would be great if you could!

I have built it locally. I forgot to mention that I also built the kernel with the same changes made for raspberry pi, but for an x86 machine this time. I wanted to see if maybe the changes was the reason. This resulted in no kernel panics. That is, compiling for x86 works fine, even when removing the ipv6 function calls and the 64-bit division cast. Maybe that is of some help.

**Comment by rapier1** [comment]
Okay, that makes sense as we've not seen these errors under x86 or 386 architectures. I should have my hands on Pi4 this Thursday. It will take me a while to figure out the build environment and get things set up. I hope to have some more information for you by early next week.

That said, if you find out anything else please let me know.

**Comment by hikikones** [comment]
Great! Thank you very much. Will let you know if I find anything else.

One thing to take note of is that USB drivers do not work on Pi4 when using mainline kernel 5.5.0. Or at least, my keyboard didn't work. But I am just using SSH so it doesn't matter. If you want keyboard support, you can use the [rpi-5.5.y](https://github.com/raspberrypi/linux/tree/rpi-5.5.y/) kernel instead, and apply [web10g patch](https://pastebin.com/raw/7jaQxXwn) onto that. The patch is missing a new line at the end due to pastebin removing it for some reason. The patch is just a diff between mainline kernel 5.5.0 and web10g 5.5.0 (but with ipv6 changes commented out).

Edit: Oh and I have also tried the rpi-5.5.y kernel with web10g patched onto it. Kernel panics also happens here.

**Comment by rapier1** [comment]
I just wanted to give you a quick update - I have a pi4 and have been going through the process of learning how to install new kernels. It's not the most intuitive process :)

Anyway, I think one of the issues is that Rasberrian compiles IPv6 as a module. This causes problem when trying to link tcp_estats into the kernel. After running the bcm2711_defconfig run 'make menuconfig' and go to 'networking options' and then change the M next to IPv6 to a *. 

While you may never need IPv6 tcp_estats assumes that IPv6 is present in the kernel. This is probably something I need to change. 

I'm still working on getting this figured out so this is just a preliminary suggestion.

**Comment by rapier1** [comment]
Okay, I was able to get things to work.
1) Ensure than IPv6 is built into the kernel and isn't compiled as a module. 
2) When you get the repo for web10-dlkm make sure you checkout the appropriate version. In this case you want to do 'git checkout kernel-4.18' and then compile. 
3) Also, when you install web10g-userland make sure that you update ld.conf.so.d to include /usr/local/lib and then run ldconfig to pick up the changes. 

Let me know if this ends up working for you. 

Also, how do you plan on using web10g in this situation? Just curious.

**Comment by hikikones** [comment]
Hello again! Just built the web10g kernel as you specified, and not a single warning or error appeared. Nice! However, I get the following error when building web10g-dlkm:

```
make -C /lib/modules/5.5.0-v7l+/build M=/home/web10g/web10g-dlkm
make[1]: Entering directory '/home/web10g/web10g-5.5'
  AR      /home/web10g/web10g-dlkm/built-in.a
  CC [M]  /home/web10g/web10g-dlkm/tcp_estats_mib_var.o
/home/web10g/web10g-dlkm/tcp_estats_mib_var.c: In function ‘read_StartTimeStamp’:
/home/web10g/web10g-dlkm/tcp_estats_mib_var.c:103:39: error: ‘struct timespec64’ has no member named ‘tv_usec’; did you mean ‘tv_sec’?
  msecs = msecs + (u64)stats->start_tv.tv_usec;
                                       ^~~~~~~
                                       tv_sec
make[2]: *** [scripts/Makefile.build:266: /home/web10g/web10g-dlkm/tcp_estats_mib_var.o] Error 1
make[1]: *** [Makefile:1693: /home/web10g/web10g-dlkm] Error 2
make[1]: Leaving directory '/home/web10g/web10g-5.5'
make: *** [Makefile:8: all] Error 2
```

I did `git checkout kernel-4.18` and then compiled with `make`. Are you using web10g kernel 4.18 for Pi4? If that is the case, how did you get bcm2711 SoC to work? I can't find any configuration for it on [rpi-4.18.y](https://github.com/raspberrypi/linux/tree/rpi-4.18.y/arch/arm/configs). To my understanding, mainline kernel for Pi4 wasn't properly supported until [version 5.5](https://www.omgubuntu.co.uk/2020/01/linux-5-5-features), and so I am using the the web10g 5.5 version.

If I compile the `kernel-5.5` version of web10g-dlkm, I get `ERROR: "__aeabi_uldivmod" [/home/web10g/web10g-dlkm/tcp_estats_nl.ko] undefined!` again, which can be "fixed" by removing the 64-bit division cast as I mentioned in the opening post. However, doing this still results in kernel panic.

Also, about point 3, did you mean `ld.so.conf` without the `d`? Since `ld.conf.so.d` is folder for me, but the folder contains a `libc.conf` file which already has the `/usr/local/lib` line. I also added the line to `ld.so.conf` and ran `ldconfig` again, but it made no difference.

I don't quite know what I'm doing wrong, considering you managed to get it working. Do you want me to write a step-by-step on my process? Maybe I am doing something silly.

> Also, how do you plan on using web10g in this situation? Just curious.

I am using web10g for my bachelor project. In short, it is about trying to optimize TCP congestion control on the sender side utilizing Active Queue Management (AQM) and Explicit Congestion Notification (ECN). It builds upon the work of our supervisor, [Alternative Backoff with ECN (ABE)](http://heim.ifi.uio.no/~naeemk/research/ABE/Networking2017ABE.pdf), which found that with ECN enabled, reducing `cwnd` by less than the response for loss (e.g. 50% with NewReno) yielded significant performance gains in lightly-multiplexed scenarios. Specifically, a 0.8 reduction factor was found to give a good latency-vs-utilization trade-off.

So the goal of the bachelor project is to try to improve upon ABE by making it more responsive to different network conditions. Instead of using a static reduction factor, it should be dynamic. And to test the results, we are using a tool called [TEACUP](http://caia.swin.edu.au/tools/teacup/) which needs web10g. :-)

**Comment by rapier1** [comment]
This was built under the most recent version of the raspberrian kernel which is 4.19. That's probably where you and I diverged on what we were trying to do. Since you are trying to build under 5.5 you'll need to double check the version of the dlkm you are building. If you type 'git tag' in that directory you'll see a tag for 'kernel-5.0'. If you check out that tag you should be able to build the dlkm successfully as that address the change to the timespec struct that was introduced in the 5.0 kernel. 

Anyway, since I was building 4.19 and you are trying for 5.5 that's where the problem arose. I'm sorry I wasn't clear on that. I included the steps I took below. Keep in mind this is for the 4.19 kernel so you'll need to double check the tags in the dlkm and userland. Ensure that you checkout the most recent kernel version (kernel-5.0 for the DLKM and kernel-4.18 for userland)

Here are the steps I took - modify them as appropriate for 5.5
1) clone the web10g repo. 
2) git checkout kis-0.13-4.19 (kis-0.13-5.5 for you)
3) git diff v4.19 > ~/web10g.patch (diff v5.5 for you)
4) clone the raspberrian kernel from https://github.com/raspberrypi/linux
5) cd into the linux directory and checkout the 4.19 tag/branch (again, 5.5y for you)
6) git apply ~/web10g.patch (this should apply successfully with minor whitespace errors)
7) then follow the directions on https://www.raspberrypi.org/documentation/linux/kernel/building.md
8) once the new kernel is up and running I cloned the web10g-dlkm, checked out the 4.18 tag and built it. (kernel-5.0 tag for you)
9) The I cloned the web10g-userland and checked out the kernel-4.18 tag, built it, and installed it
10) then I added a file into ld.so.conf.d called 'libestats.conf' which contained the line '/usr/local/lib' and then ran ldconfig to pick up the new library location (where libestats lives)

Also, I'm really interested in the project. I do some work with Dr. Justine Sherry at CMU and she's doing a lot of network research like this. Once you get to the point where you have something to share I'd really like to see it.

**Comment by hikikones** [comment]
Ah I see! Well, I have now done the same as you and also with the same kernel (4.19). It doesn't really matter which kernel I use, as long as it works. No errors occurred at any stage when building. I can load web10g fine with `modprobe tcp_estats_nl` and start logging with `sysctl -w net.ipv4.tcp_estats=127`. Listing connections with `web10g-listconns` also works. I can also start up a TCP client/server on localhost with `netcat` and see all the new TCP statistics from web10g. However, I sadly still get kernel panic if I now `ssh` into the web10g enabled computer from another computer on the network.

Have you tried to `ssh` into the web10g enabled computer from another computer on the network, while it is logging? If it works for you, I will need to retrace my step, although I don't quite know what to look for.

Also, I did not know about the `git diff v4.19` command. What a handy command. I used to either manually create a patch by looking at your web10g changes, or take the diff between mainline kernel and your modified version, and then clean it up.

> Also, I'm really interested in the project. I do some work with Dr. Justine Sherry at CMU and she's doing a lot of network research like this. Once you get to the point where you have something to share I'd really like to see it.

Of course! Will let you know once it is ready. But don't expect too much of it. :-)

Oh and thanks for the help so far. I greatly appreciate it.
